pkg/operator/configobserver/node/observe_latency_profile.go

```diff
@@ -56,42 +53,37 @@ func (l *latencyProfileObserver) observeLatencyProfile(
 	existingConfig map[string]interface{},
 ) (ret map[string]interface{}, errs []error) {
 
-	// traverse list of latencyConfigs to get all the config paths
-	configPaths := make([][]string, len(l.latencyConfigs))
-	for i, latencyConfig := range l.latencyConfigs {
-		configPaths[i] = latencyConfig.ConfigPath
-	}
-
 	defer func() {
 		// Prune the observed config so that it only contains fields specific to this observer
-		ret = configobserver.Pruned(ret, configPaths...)
+		ret = configobserver.Pruned(ret, l.usedConfigPaths...)
 	}()
 
+	for _, shouldSupressConfigUpdatesFn := range l.shouldSuppressConfigUpdatesFuncs {
+		suppress, reason, err := shouldSupressConfigUpdatesFn()
+		if err != nil {
+			errs = append(errs, err)
+			return existingConfig, errs
+		}
+		if suppress {
+			// log that latency profile couldn't be updated due to conditional
+			klog.Warningf("latency profile config observer suppressed update to observed config: %s", reason)
+			return existingConfig, errs
+		}
+	}
+
 	listers := genericListers.(NodeLister)
 	configNode, err := listers.NodeLister().Get("cluster")
 	// we got an error so without the node object we are not able to determine worker latency profile
 	if err != nil {
 		// if config/v1/node/cluster object is not found, that can be treated as a non-error case
 		if !apierrors.IsNotFound(err) {
 			errs = append(errs, err)
+		} else { // but raise a warning
+			klog.Warningf("nodes.config.openshift.io/cluster object could not be found")
 		}
 		return existingConfig, errs
 	}
 
-	// if condition is not satisfied, return existing config back without any changes
-	// eg. shouldSuppressConfigUpdatesFunc is to check if rollouts are in progressing state
-	// then reject the update and do not set any newly observed values till the last rollout is complete
-	suppress, err := l.shouldSuppressConfigUpdatesFunc()
-	if err != nil {
-		errs = append(errs, err)
-		return existingConfig, errs
-	}
-	if suppress {
-		// log that latency profile couldn't be updated due to conditional
-		klog.Warningf("latency profile config observer did not update observed config due to unsatisified pre-requisite")
-		return existingConfig, errs
-	}
-
 	if configNode.Spec.WorkerLatencyProfile == "" {
```

Comment:
after playing with this, this seems like a bug. I acknowledge that we do not have to update the status since the feature will be "disabled", but here we should at least return the value back to its original one.

if you do `""` -> `"MediumUpdateAverageReaction"` -> `""`

the KCM, etc will keep using the last set value (in this case `MediumUpdateAverageReaction`) and will never revert back to the default behaviour. In case of KCM from `2m` to default `40s`.

This is unexpected IMO and also you cannot read the configs to asses the actual state of the cluster

Reply:
Yeah, since `existingConfig` is being returned when we switch back to `""` (empty) profile, it's a bug of not rightly having default profile related args during this state.

However, the workaround in my mind is to use the following inside `if configNode.Spec.WorkerLatencyProfile == "" { ... }`:
- `return nil, errs` or
- maybe `return map[string]interface{}{}, errs`

essentially both will have same effect, so we can use either.

That way this observer would return a blank config (instead of previous/existing config) when empty profile is reactivated on the cluster.

Ref:
https://github.com/openshift/library-go/blob/8b2df431789c3ba021555d3cc5029ee15021dbc3/pkg/operator/configobserver/config_observer_controller.go#L136-L149

The config observer (&/ mergo [1]) would be able to merge the `nil`/blank config giving us an observedConfig that do not contain the latency profile related args anymore at the end of the config observation sync loop (reconciliation). This would I guess also help keep the observer in a some what no-op state (state where observedConfig is untouched) when latency profile set on the cluster is empty.

[1] https://github.com/openshift/library-go/blob/master/vendor/github.com/imdario/mergo/merge.go#L208-L210, https://github.com/openshift/library-go/blob/master/vendor/github.com/imdario/mergo/merge.go#L236-L255, https://github.com/openshift/library-go/blob/master/vendor/github.com/imdario/mergo/mergo.go#L62-L65

Reply:
The reasoning behind return blank config values for empty profile instead of the default arg val pairs directly would be in favour of cluster downgrades/upgrades and not break them somehow.

Some old context: https://github.com/openshift/cluster-kube-apiserver-operator/pull/1328#discussion_r855272795 that could be helpful here.

Reply:
@atiratree since this is buggy, we have to fix KASO as well. Bumping library-go in KASO after merging this PR should suffice.
cc: @p0lyn0mial, already have [a PR for this](https://github.com/openshift/cluster-kube-apiserver-operator/pull/1360).

Reply:
Cool.
1. I have updated it to use `map[string]interface{}{}`.
2. Added the comment in place explaining it
3. Seems like we already have the test cases for that in Scenario 1 of `TestCreateLatencyProfileObserverKAS` and `TestCreateLatencyProfileObserverKCM`.

Reply:
Added the following test cases as well in `TestCreateLatencyProfileObserverKAS` and `TestCreateLatencyProfileObserverKCM` to check while going from Medium/Default -> "" profile.
```go
                // scenario 6: Update worker latency profile from MediumUpdateAverageReaction to Empty
		{
			name: "worker latency profile update from MediumUpdateAverageReaction to Empty profile: config with empty values",
			existingConfig: map[string]interface{}{
				"extendedArguments": map[string]interface{}{
					"node-monitor-grace-period": []interface{}{"2m0s"},
				},
			},
			expectedObservedConfig: map[string]interface{}{},
			workerLatencyProfile:   "",
		},

		// scenario 7: Update worker latency profile from Default to Empty
		{
			name: "worker latency profile update from Default to Empty profile: config with empty values",
			existingConfig: map[string]interface{}{
				"extendedArguments": map[string]interface{}{
					"node-monitor-grace-period": []interface{}{"40s"},
				},
			},
			expectedObservedConfig: map[string]interface{}{},
			workerLatencyProfile:   "",
		},
```